prow/gerrit/adapter/adapter.go

```diff
@@ -222,44 +223,76 @@ func createCache(cloneURI *url.URL, cookieFilePath string, cacheSize int, config
-		log := logrus.WithField("host", instance)
-		for _, change := range changes {
-			log := log.WithFields(logrus.Fields{
-				"branch":   change.Branch,
-				"change":   change.Number,
-				"repo":     change.Project,
-				"revision": change.CurrentRevision,
-			})
+		wg.Add(1)
+		go func(instance string, changes []gerrit.ChangeInfo) {
+			defer wg.Done()
+			log := logrus.WithField("host", instance)
+			for _, change := range changes {
+				log := log.WithFields(logrus.Fields{
+					"branch":   change.Branch,
+					"change":   change.Number,
+					"repo":     change.Project,
+					"revision": change.CurrentRevision,
+				})
 
-			cloneURI, err := makeCloneURI(instance, change.Project)
-			if err != nil {
-				return fmt.Errorf("makeCloneURI: %w", err)
-			}
+				cloneURI, err := makeCloneURI(instance, change.Project)
+				if err != nil {
+					errChan <- fmt.Errorf("makeCloneURI: %w", err)
+					return
+				}
 
-			cache, ok := c.repoCacheMap[cloneURI.String()]
-			if !ok {
-				if cache, err = createCache(cloneURI, c.cookieFilePath, c.cacheSize, c.configAgent); err != nil {
-					return err
+				repoCacheMapMux.Lock()
+				cache, ok := c.repoCacheMap[cloneURI.String()]
+				if !ok {
+					if cache, err = createCache(cloneURI, c.cookieFilePath, c.cacheSize, c.configAgent); err != nil {
+						repoCacheMapMux.Unlock()
+						errChan <- err
+						return
+					}
+					c.repoCacheMap[cloneURI.String()] = cache
 				}
-				c.repoCacheMap[cloneURI.String()] = cache
-			}
+				repoCacheMapMux.Unlock()
 
-			result := client.ResultSuccess
-			if err := c.processChange(log, instance, change, cloneURI, cache); err != nil {
-				result = client.ResultError
-				log.WithError(err).Info("Failed to process change")
+				result := client.ResultSuccess
+				if err := c.processChange(log, instance, change, cloneURI, cache); err != nil {
+					result = client.ResultError
+					log.WithError(err).Info("Failed to process change")
+				}
+				prometheusMux.Lock()
+				gerritMetrics.processingResults.WithLabelValues(instance, change.Project, result).Inc()
+				prometheusMux.Unlock()
+
+				latestMux.Lock()
+				lastTime, ok := latest[instance][change.Project]
+				if !ok || lastTime.Before(change.Updated.Time) {
+					lastTime = change.Updated.Time
+					latest[instance][change.Project] = lastTime
+				}
+				latestMux.Unlock()
 			}
-			gerritMetrics.processingResults.WithLabelValues(instance, change.Project, result).Inc()
+			log.Infof("Processed %d changes", len(changes))
+		}(instance, changes)
+	}
 
-			lastTime, ok := latest[instance][change.Project]
-			if !ok || lastTime.Before(change.Updated.Time) {
-				lastTime = change.Updated.Time
-				latest[instance][change.Project] = lastTime
-			}
+	// Capture errors here so that goroutines are unblocked.
+	var errs []error
+	go func() {
+		for err := range errChan {
+			errs = append(errs, err)
 		}
-		log.Infof("Processed %d changes", len(changes))
+	}()
+	wg.Wait()
+	close(errChan)
+
+	if len(errs) > 0 {
```

Comment:
nit: You don't need to do this check. `utilerrors.NewAggregate(errs)` is smart enough to both ignore `nil` errors and return nil if there are no errors.

Reply:
Oops yeah I actually saw that after commenting this then forgot to delete this comment. :man_facepalming: